src/torrent.cpp

```diff
@@ -3804,13 +3806,10 @@ namespace libtorrent
 		// since this piece just passed, we might have
 		// become uninterested in some peers where this
 		// was the last piece we were interested in
-		for (peer_iterator i = m_connections.begin();
-			i != m_connections.end();)
+		// update_interest may disconnect the peer and
+		// invalidate the iterator
+		for (auto p : clone(m_connections))
 		{
-			peer_connection* p = *i;
```

Comment:
here, `->self()` is not necessary (I think), because `p->update_interest();` is at the end of the loop